blockstore/apps/bundles/store.py

```diff
@@ -279,27 +319,184 @@ def _create_snapshot(self, bundle_uuid, files):
-        files[str(path)] = self._save_file(snapshot.bundle_uuid, path, data, public)
-        return self._create_snapshot(snapshot.bundle_uuid, files)
+        def file_for_path(path):
+            if path in draft.files_to_overwrite:
+                storage_path = self._data_file_path(draft.bundle_uuid, draft.name, path)
+                return self.open(draft, storage_path)
+            return self.snapshot_store.open(draft.base_snapshot, path)
+
+        paths_to_files = {
+            path: file_for_path(path)
+            for path in draft.files
+        }
+        try:
+            yield paths_to_files
+        finally:
+            for open_file in paths_to_files.values():
+                open_file.close()
 
-    def snapshot_by_removing_path(self, snapshot, path):
+    def update_files(self, bundle_uuid: UUID, name:str, updated_files:dict, updated_at=None) -> Draft:
         """
-        Create and save a BundleSnapshot with file at path removed from snapshot.
+        Update the draft with some number of files.
+
+        `updated_files` is a dict with paths for keys and Files for values.
         """
-        files = {p: snapshot.files[p] for p in snapshot.files if p != path}
-        return self._create_snapshot(snapshot.bundle_uuid, files)
+        # First write all the files out in their appropriate draft space.
+        new_files_written = {}
+        existing_draft = self.get(bundle_uuid, name)
+        for path, django_file in updated_files.items():
+            storage_path = self._data_file_path(bundle_uuid, name, path)
+            file_info = FileInfo(
+                path=path,
+                public=False,  # Hardcoded for now -- maybe get rid of this from FileInfo altogether?
+                size=django_file.size,
+                hash_digest=FileInfo.generate_hash(django_file).digest()
+            )
+            # TODO: Add check to remove file from overwrites if we've overwritten
+            # the file to be the same as it is in the base_snapshot.
+            if default_storage.exists(storage_path):
+                default_storage.delete(storage_path)
+            default_storage.save(storage_path, django_file)
+            new_files_written[path] = file_info
+
+        new_draft = Draft(
+            bundle_uuid=bundle_uuid,
+            name=name,
+            base_snapshot=existing_draft.base_snapshot,
+            files_to_overwrite={**existing_draft.files_to_overwrite, **new_files_written},
+            created_at=existing_draft.created_at,
+            updated_at=updated_at or datetime.now(timezone.utc)
+        )
+        draft_summary_path = self._summary_path(bundle_uuid, name)
+        draft_summary_json = json.dumps(new_draft, cls=BundleDataJSONEncoder, indent=2, sort_keys=True)
+        if default_storage.exists(draft_summary_path):
+            default_storage.delete(draft_summary_path)
+        default_storage.save(draft_summary_path, ContentFile(draft_summary_json))
```

Comment:
This uses `delete` followed immediately by `save`. I think S3 uses an eventual consistency model for deletes, and if so, is it possible the `save()` might still see the deleted file and then ["modify the filename as necessary to get a unique name"](https://docs.djangoproject.com/en/2.1/ref/files/storage/#django.core.files.storage.Storage.save) ?

Reply:
So that used to be an issue with US-East in particular, but even they've upgraded so that read-after-write consistency shouldn't be a problem. You're right that the reason I have the delete/save silliness is that `save` won't overwrite by default. I tried using basic `open` functionality but Django's storages will only open something in binary mode (it seems to ignore binary/string mode parameter and always open in binary), which causes issues with `json.dump`. Anyhow, I'll fix it so it's one atomic operation.

Reply:
I think it's also worth mentioning that some django-storages backends have an option to allow direct overwrites via `save`, e.g. https://github.com/jschneier/django-storages/search?q=file_overwrite&unscoped_q=file_overwrite